### PR TITLE
SALTO-3143: Deploy config for missing users & handle missing users in Zendesk

### PR DESCRIPTION
**Pull request by shir-reifenberg**
Add deploy config option + handle missing users in Zendesk

---

Changes:

Adapter-components

- Add 'defaultMissingUserFallback' config option for fallback when users referenced from elements are missing

Zendesk
- Add deploy config to zendesk by default, and add this config as an optional param for filters & change validator
- Adjust `missingUsersValidator` to notify users in case missing users values are about to change
- Add filter that on `onDeploy` will change all missing users in changed elements to the user provided in deploy config

---
_Release Notes_: 

Zendesk
- Add the option to define fallback username for elements with references to missing users.

---
_User Notifications_: 
None

### CODE REVIEW

**Comment by alonstern** [comment]
@shir-reifenberg Can we separate the adapter-component changes to a different PR to unblock the Jira task for handling missing users?

File: packages/zendesk-adapter/src/user_utils.ts [hunk 1]
```diff
@@ -170,3 +188,33 @@ const getUsersFunc = ():(paginator: clientUtils.Paginator) => Promise<User[]> =>
...
+): Promise<string | undefined> => {
+  if (defaultMissingUserFallback === configUtils.DEPLOYER_FALLBACK_VALUE) {
+    try {
+      const response = (await client.getSinglePage({
+        url: '/api/v2/users/me',
+      })).data
+      if (isCurrentUserResponse(response)) {
+        return response.user.email
+      }
+      return undefined
```
Comment (shir-reifenberg) [inline packages/zendesk-adapter/src/user_utils.ts:209]: @netama WDYT we should do if `isCurrentUserResponse` is not valid?

File: packages/adapter-components/src/config/shared.ts [hunk 2]
```diff
@@ -221,3 +221,20 @@ export const validateSupportedTypes = (
     throw Error(`Invalid type names in ${fetchConfigPath}: ${invalidIncludedTypes} does not match any of the supported types.`)
   }
 }
+
+/**
+ * Verify defaultMissingUserFallback value in deployConfig is valid
+ */
+export const validateDeployConfig = (
```
Comment (netama) [inline packages/adapter-components/src/config/shared.ts:228]: 👌

File: packages/zendesk-adapter/src/config.ts [hunk 3]
```diff
@@ -2582,6 +2586,9 @@ export const configType = createMatchingObjectType<Partial<ZendeskConfig>>({
         },
       ),
     },
+    [DEPLOY_CONFIG]: {
```
Comment (netama) [inline packages/zendesk-adapter/src/config.ts:2589]: (not for this line) do we want to omit this section by default or not?

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 4]
```diff
@@ -38,11 +40,62 @@ const getMissingUsers = (instance: InstanceElement, existingUsers: Set<string>):
...
+    elemID: instance.elemID,
+    severity: 'Error',
+    message: MISSING_USERS_ERROR_MSG,
+    detailedMessage: `The following users are referenced by this element, but do not exist in the target environment: ${missingUsers.join(', ')}.\nIn addition, we could not get the defined fallback user ${userFallbackValue}. In order to deploy this element, add these users to your target environment, edit this element to use valid usernames, or set the target environment's user fallback options. Learn more: ${MISSING_USERS_DOC_LINK}`,
+  }
+}
+
 /**
- * Verifies users exists before deployment of an element with user fields
+ * Verifies users exists before deployment of an element with user fields.
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:94]: nit: `users exists` -> `users exist` / `user exists`

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 5]
```diff
@@ -38,11 +40,62 @@ const getMissingUsers = (instance: InstanceElement, existingUsers: Set<string>):
...
+ * If the deploy option 'defaultMissingUserFallback' exists, warns the user about missing users changes
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:95]: nit: maybe explain here what the behavior is (summarize the cases briefly)?

File: packages/zendesk-adapter/src/user_utils.ts [hunk 6]
```diff
@@ -16,32 +16,50 @@
...
+import ZendeskClient from './client/client'
 import { ValueReplacer, replaceConditionsAndActionsCreator, fieldReplacer } from './replacers_utils'
 
 const log = logger(module)
 const { toArrayAsync } = collections.asynciterable
 const { makeArray } = collections.array
 
-export type User = {
+const MISSING_DEPLOY_CONFIG_USER = 'User provided in defaultMissingUserFallback does not exist in the target environemt'
+// system options that does not contain a specific user value
```
Comment (netama) [inline packages/zendesk-adapter/src/user_utils.ts:31]: nit (I know this was just moved): `does` -> `do`

File: packages/zendesk-adapter/src/user_utils.ts (hunk 1 above)
Comment (netama) [reply packages/zendesk-adapter/src/user_utils.ts:209]: I think the fallback in the `catch` below should apply for this as well (let's just make sure our logs differentiate between them, but the user-facing part can be the same IMO)

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 7]
```diff
@@ -38,11 +40,62 @@ const getMissingUsers = (instance: InstanceElement, existingUsers: Set<string>):
...
+): ChangeError | undefined => {
+  const missingUsers = getMissingUsers(instance, userEmails)
+  if (_.isEmpty(missingUsers)) {
+    return undefined
+  }
+  return {
+    elemID: instance.elemID,
+    severity: 'Error',
+    message: MISSING_USERS_ERROR_MSG,
+    detailedMessage: `The following users are referenced by this element, but do not exist in the target environment: ${missingUsers.join(', ')}.\nIn order to deploy this element, add these users to your target environment, edit this element to use valid usernames, or set the target environment's user fallback options. Learn more: ${MISSING_USERS_DOC_LINK}`,
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:55]: I don't think this should say `this element`, can we follow the usual pattern of e.g. `${type} instance ${name}`?

File: packages/zendesk-adapter/test/change_validators/missing_users.test.ts [hunk 8]
```diff
@@ -83,29 +86,94 @@ describe('missingUsersValidator', () => {
...
-        detailedMessage: 'The following users don\'t exist in the target environment: thisuserismissing@salto.com, thisuserismissing2@salto.com.\nPlease manually edit the element and set existing user emails or add users with this emails to the target environment.',
+        message: 'Element references users which don\'t exist in target environment',
+        detailedMessage: 'The following users are referenced by this element, but do not exist in the target environment: thisuserismissing@salto.com, thisuserismissing2@salto.com.\nIn order to deploy this element, add these users to your target environment, edit this element to use valid usernames, or set the target environment\'s user fallback options. Learn more: https://help.salto.io/en/articles/6955302-element-references-users-which-don-t-exist-in-target-environment-zendesk',
       },
     ])
   })
-  it('should not return a warning if user exists', async () => {
+  it('should return warning in case of overriding missing users with user from deploy config', async () => {
+    const changes = [toChange({ after: articleInstance }), toChange({ after: macroInstance })]
+    deployConfig = { defaultMissingUserFallback: '4@4' }
```
Comment (netama) [inline packages/zendesk-adapter/test/change_validators/missing_users.test.ts:102]: nit: I'm not 100% sure this is _not_ a valid email, but I think we shouldn't have it as the example in the test cases (if you do want to verify it exists let's have it in one case)

File: packages/zendesk-adapter/src/change_validators/missing_users.ts [hunk 9]
```diff
@@ -15,19 +15,21 @@
...
+import { ZedneskDeployConfig } from '../config'
 
 const { awu } = collections.asynciterable
 const { createPaginator } = clientUtils
-// system options that does not contain a specific user value
-const VALID_USER_VALUES = ['current_user', 'all_agents', 'requester_id', 'assignee_id', 'requester_and_ccs', 'agent', 'end_user', '']
+const { isDefined } = lowerdashValues
+
+const MISSING_USERS_DOC_LINK = 'https://help.salto.io/en/articles/6955302-element-references-users-which-don-t-exist-in-target-environment-zendesk'
+const MISSING_USERS_ERROR_MSG = 'Element references users which don\'t exist in target environment'
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/missing_users.ts:32]: nit: I think we usually say `instance`, is there a reason to say `element` here?

File: packages/zendesk-adapter/test/change_validators/missing_users.test.ts [hunk 10]
```diff
@@ -83,29 +86,94 @@ describe('missingUsersValidator', () => {
       {
         elemID: articleInstance.elemID,
         severity: 'Error',
-        message: '1 references to users that don\'t exist in the target environment.',
-        detailedMessage: 'The following users don\'t exist in the target environment: article@salto.com.\nPlease manually edit the element and set existing user emails or add users with this emails to the target environment.',
+        message: 'Element references users which don\'t exist in target environment',
+        detailedMessage: 'The following users are referenced by this element, but do not exist in the target environment: article@salto.com.\nIn order to deploy this element, add these users to your target environment, edit this element to use valid usernames, or set the target environment\'s user fallback options. Learn more: https://help.salto.io/en/articles/6955302-element-references-users-which-don-t-exist-in-target-environment-zendesk',
```
Comment (netama) [inline packages/zendesk-adapter/test/change_validators/missing_users.test.ts:90]: nit: do we want a newline before the "learn more"?

**Review by netama — APPROVED** [review]
looks great, added some nits but LGTM otherwise

File: packages/zendesk-adapter/test/change_validators/missing_users.test.ts (hunk 10 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/test/change_validators/missing_users.test.ts:90]: I thought it looks better as the link is very long

File: packages/zendesk-adapter/src/change_validators/missing_users.ts (hunk 7 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/src/change_validators/missing_users.ts:55]: Maybe just instance? According to @liorn the instance name and type already appear in the change error table (in the ui) so it's redundant

File: packages/zendesk-adapter/src/config.ts (hunk 3 above)
Comment (shir-reifenberg) [reply packages/zendesk-adapter/src/config.ts:2589]: Yep

**Pull request merged**